### ISSUE TITLE
Include BOM handling in ANTLRInputStream

### ISSUE THREAD

**Issue by mike-lischke** [tunnelvisionlabs/antlr4ts#319]
It would be helpful if the ANTLRInputStream class could deal with a potential BOM at the beginning of the input (e.g. when loaded from a file).

**Comment by sharwell** [comment]
ANTLRInputStream in the TypeScript target does not operate at this level. The input is expected to already be converted to a string based on the proper encoding. If a byte order mark remains at this time, it is assumed that the byte order mark really is a U+FFFE character in the input.

**Comment by BurtHarris** [comment]
Agreed in purpose! However I'm not convinced I'd do it on that class.      

This is very compatible with an approach I've been considering by taking on a dependency on `vinyl` which seems to do a great job of abstracting files and file metadata.     Together with Vinyl adapter like `vinyl-fs` , it really seems to hit the spot.

**Comment by sharwell** [comment]
:memo: Note that I am generally in favor of byte order mark handling, and filed/fixed a similar issue in the ANTLR tool for 4.2: antlr/antlr4#175

**Comment by BurtHarris** [comment]
As background, I ran into Vinyl in researching build systems.   I've spent too much time fooling around with the package.js scripts, and wanted something smarter.   Jake and Grunt both seem really crusty, but [Gulp](http://gulpjs.com/) seems based on modern node.js and ECMAScript v-next concepts like streams.

**Comment by mike-lischke** [comment]
I find it convenient to not have to tinker with string input at this point and hence have implemented BOM handling in the C++ target (in ANTLRInputStream). Would be nice if other tagets would do the same.

**Comment by sharwell** [comment]
@mike-lischke The C# target also implements BOM handling in ANTLRInputStream. The difference with the TypeScript target is it's the only one that doesn't do *anything* with encoding because it doesn't read raw streams. There is only one supported input form, which is whatever JavaScript uses for the implementation of the `string` type.

**Comment by mike-lischke** [comment]
Well the BOM is not so much for handling raw streams, which means you can remove it also from strings (otherwise a user of the input stream class has to test this for every source of input (file, clipboard, editors etc.)). I don't see a problem by removing it. It's an otherwise invalid code sequence and shouldn't be used anywhere else anyway.

**Comment by sharwell** [comment]
As an example, this would break cases where someone is parsing content similar to the serialized ATN we create for ANTLR 4. Obviously we can't remove all U+FFFE characters, but even if we remove one from the start of an already-decoded input we're changing parsing behavior in ways that can break applications.

**Comment by sharwell** [comment]
@mike-lischke I'm reopening this as a documentation issue. Since the TypeScript target differs from ANTLR in this respect, I believe users would find it valuable if the documentation contained an example related to the parsing of inputs which are encoded as UTF-8 with a byte order mark.

**Comment by BurtHarris** [comment]
`string`s are no place for binary data, in Node.js we use `Buffer` for that.   If the input to a reader is a `string`, striping any leading BOM seems perfectly safe to me.
